numba_dpex/core/kernel_interface/dispatcher.py

```diff
@@ -626,12 +632,25 @@ def __call__(self, *args):
                     cache=self._cache,
                 )
 
-        # create a sycl::KernelBundle
-        kernel_bundle = dpctl_prog.create_program_from_spirv(
-            exec_queue,
-            device_driver_ir_module,
-            " ".join(self._create_sycl_kernel_bundle_flags),
+        kernel_bundle_key = build_key(
+            tuple(argtypes),
+            self.pyfunc,
+            dpex_kernel_target.target_context.codegen(),
+            backend=backend,
```

Comment:
The key for the `kernel_bundle` has to store the queue object and not `backend` and `device_type`. A kernel bundle is tied to a SYCL context and having the same device_type and backend does not guarantee that the context is the same. If we pass a kernel_bundle that was compiled for a specific context to a queue on a different context that will lead to a crash.

https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_the_kernel_bundle_class

Reply:
Regarding the point raised by [fcharras](https://github.com/fcharras), do we really need to hash the codegen as part of the cache key? It may be adding some extra overhead.